### PR TITLE
[IA-1986] Update attached PD when updating runtime

### PR DESCRIPTION
**Pull request by akarukappadath**
1. Fix a bug in existing update disk size where we're updating boot disk size
2. Fix a bug where update disk size doesn't work because runtime is not restarted.
3. Add automation tests to validate update disk size works for gce, gce with pd, dataproc

---
Have you read [CONTRIBUTING.md](https://github.com/DataBiosphere/leonardo/blob/develop/CONTRIBUTING.md) lately? If not, do that first.

I, the developer opening this PR, do solemnly pinky swear that:

- [ ] I've documented my API changes in Swagger

In all cases:

- [ ] Get a thumbsworth of review and PO signoff if necessary
- [ ] Verify all tests go green
- [ ] Squash and merge; Delete your branch after this
- [ ] Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

**Comment by akarukappadath** [comment]
jenkins retest

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/RuntimeServiceInterp.scala
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/RuntimeServiceInterp.scala]: can you make this into a function so that it can be reused in `processUpdateGceConfigRequest`

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/RuntimeServiceInterp.scala
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/RuntimeServiceInterp.scala]: while you're in here, can you test to see if disk size change require restart? (there was a bug ticket that updating disk size didn't take effect until machine is restarted

Comment (akarukappadath) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/RuntimeServiceInterp.scala]: can you elaborate on the bug? I just tried updating a disk size and it looks like it went through fine, but was the machine unable to use the disk?

Comment (Qi77Qi) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/RuntimeServiceInterp.scala]: https://broadworkbench.atlassian.net/browse/IA-2015 is the ticket...

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/BaseCloudServiceRuntimeMonitor.scala
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/BaseCloudServiceRuntimeMonitor.scala]: why is this needed?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala]: we don't have to run this async to `restartRuntime`....reason why we used asyncTaskProcessor in a main message handler is because we want this whole waiting happen async to message acknowledgement (we don't want to wait say 20 min before acking a message, cuz things get weird with pubsub in that case)...

Here the whole pollOperation is already async to the message acking, so you don't need to run ilne 556 async to `restartRuntime` anymore

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala]: is this to account for when a `updateDisk` api called..if the disk is attached to a runtime, then we try to stop the attached runtime if it happens to be running? in other words, `restartRuntimeId` only exists when there's a runtime associated with the disk, and the runtime happens to be running?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/RuntimeServiceInterp.scala
Comment (Qi77Qi) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/RuntimeServiceInterp.scala]: why don't we merge this intent in `UpdateRuntimeMessage` instead of a separate message for updating diskMesage? won't be get in race condition this way since `UpdateRuntimeMessage` can potentially trigger a restart as well?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/BaseCloudServiceRuntimeMonitor.scala
Comment (akarukappadath) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/BaseCloudServiceRuntimeMonitor.scala]: once a runtime is ready (either the init or start-up script is complete) the disk will have been resized

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala
Comment (akarukappadath) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala]: `restartRuntimeId` only exists if the `UpdateDiskMessage` came from a `updateRuntime` request. If it's coming from a `updateDisk` request, `restartRuntimeId` will be `None`. The reasoning behind this is because we have the `allowStop` field in `updateRuntime,` but it doesn't make sense for the `updateDisk` endpoint. So, if someone updates the disk size through the `updateRuntime` endpoint it will restart the runtime for the user if `allowStop` is true (or throw an error if `allowStop` is false). If someone updates the disk size through the `updateDisk` endpoint, the runtime will not be restarted but the disk will have `needsResize=true` until they restart the runtime themselves

**Comment by akarukappadath** [comment]
jenkins retest

File: http/src/main/resources/init-resources/startup.sh
Comment (Qi77Qi) [inline http/src/main/resources/init-resources/startup.sh:160]: this does nothing if the disk doesn't need resize...Azmina originally added `needsResize` variable, but it was added to persistent disk table, which is not the right place because we need this flag for non pd gce runtimes too. I removed `needsResize` entirely since it doesn't seem to me worth the extra complexity.

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/runtimes/RuntimePatchSpec.scala
Comment (rtitle) [inline automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/runtimes/RuntimePatchSpec.scala:95]: Really.. we specify `DiskSize(20)` and get `22G` from `df -H`?

File: http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala
Comment (rtitle) [inline http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala:370]: Curious why we need this instead of `getActiveClusterByNameMinimal` which seems to be equivalent.. our usual practice is to not return `*Record` objects outside of DB code

Comment (Qi77Qi) [reply http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/db/ClusterComponent.scala:370]: I added this because originally we were using `getActiveClusterByNameMinimal`, which stubs out fields outside cluster table. I didn't want to keep using `getActiveClusterByNameMinimal` because this caused a bug where I was using dataproc instances from `Runtime` object, but because it was stubbed out, it wasn't doing what I expected it's doing. By using  `getActiveClusterRecordByName`, the return type is exactly what it is, and there's no chance that caller might accidentally be using a stubbed out field

File: automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/runtimes/RuntimePatchSpec.scala
Comment (Qi77Qi) [reply automation/src/test/scala/org/broadinstitute/dsde/workbench/leonardo/runtimes/RuntimePatchSpec.scala:95]: yea...not sure why

**Review by rtitle — APPROVED** [review]
LGTM after merge conflicts are addressed

**Pull request merged**
